crates/biome_js_analyze/src/lint/nursery/use_sorted_classes/tailwind_preset.rs

```diff
@@ -572,11 +573,684 @@ const UTILITIES_LAYER_CLASSES: [&str; 569] = [
     "duration-",
     "ease-",
     "will-change-",
+    "contain-none$",
+    "contain-content$",
+    "contain-strict$",
+    "contain-size$",
+    "contain-inline-size$",
+    "contain-layout$",
+    "contain-paint$",
+    "contain-style$",
     "content-",
     "forced-color-adjust-auto$",
     "forced-color-adjust-none$",
 ];
 
+pub fn get_variant_classes() -> Vec<Variant> {
```

Comment:
Pinging @ematipico for perf advice here. What is needed is a map between strings that identify variants and their respective "weights". A weight is a bitvec, and the way it works is the following.

Variants exist in a specific order. E.g. `var1`, `var2` and `var3`.

For each entry, we need to compute a weight which is a bit vector of `1` shifted `n` times, `n` being the index in the ordered list I mentioned. So in our example, that'd be:

- `var1`: `1`
- `var2`: `10`
- `var3`: `100`

And so on. Tailwind classes can have multiple variants, e.g. `var1:var3:m-4`. When that happens, those bit vectors are xor'd, like this: `1` xor `100` = `101`. With the resulting bit vector, classes are sorted.

I would love your input in how to optimize this. We can either just find a way to reduce computation, or memory consuption, etc - or we could even figure out some smart shortcut that has the same result as this xor thing but better perfomance. Any ideas?

Reply:
> Pinging @ematipico for perf advice here. What is needed is a map between strings that identify variants and their respective "weights". A weight is a bitvec, and the way it works is the following.
> 
> Variants exist in a specific order. E.g. `var1`, `var2` and `var3`.
> 
> For each entry, we need to compute a weight which is a bit vector of `1` shifted `n` times, `n` being the index in the ordered list I mentioned. So in our example, that'd be:
> 
> * `var1`: `1`
> * `var2`: `10`
> * `var3`: `100`
> 
> And so on. Tailwind classes can have multiple variants, e.g. `var1:var3:m-4`. When that happens, those bit vectors are xor'd, like this: `1` xor `100` = `101`. With the resulting bit vector, classes are sorted.
> 
> I would love your input in how to optimize this. We can either just find a way to reduce computation, or memory consuption, etc - or we could even figure out some smart shortcut that has the same result as this xor thing but better perfomance. Any ideas?

I don't know if this changes the math a bit, but just want to point out here too that I've seen that `n` as the times `1` is shifted isn't strictly the index of the generated array of variants.
It will probably work well for single variants classes, but when it comes to doing bitwise XOR, I think the result would change.
Haven't tested it out yet and I'm of course interested in any way that could make this more performant 😄

Reply:
Are there limitations on the number of weights that a variant can store? For example, is it realistic to have 1000 variants? Having 1000 variants would mean that our highest base weight would be `1000*100`, plus the singular variations.

Also, do we expect `get_variant_classes` to receive variables? If not, we should make sure to create these variants only once. This is achievable with [`OnceCell`](https://doc.rust-lang.org/std/cell/struct.OnceCell.html) or with `lazy_static!`, depending on the use case. Use `OnceCell` if the function needs to accept arguments, use `lazy_static!` otherwise.

Another piece of optimisation. Use `smallvec` instead of a `Vector`. The vector tends to save info inside the heap, but **if you know** the number of the variants like in this case, `smallvec` will save the data in the stack.

Reply:
> Are there limitations on the number of weights that a variant can store? For example, is it realistic to have 1000 variants? Having 1000 variants would mean that our highest base weight would be `1000*100`, plus the singular variations.
> 
> Also, do we expect `get_variant_classes` to receive variables? If not, we should make sure to create these variants only once. This is achievable with [`OnceCell`](https://doc.rust-lang.org/std/cell/struct.OnceCell.html) or with `lazy_static!`, depending on the use case. Use `OnceCell` if the function needs to accept arguments, use `lazy_static!` otherwise.

Each variant has its own weight, which right now has a max at a BitVec with `166` elements inside for the `print` variant.
Tailwind accepts user-defined variants and those will be put after the default ones unless you override the whole default order configuration: [Custom Variant Ordering](https://v2.tailwindcss.com/docs/configuring-variants#ordering-custom-variants)

I haven't dug deeper on user-defined variants and their ordering and thus is not included in this PR, but it's safe to assume there will be no limit to the number of variants.

> I don't know if this changes the math a bit, but just want to point out here too that I've seen that n as the times 1 is shifted isn't strictly the index of the generated array of variants.
> It will probably work well for single variants classes, but when it comes to doing bitwise XOR, I think the result would change.

I also have to retract this in some way, I've noticed two jumps on `weight`:
- between `marker` and `selection`
- between `selection` and `file`
and that should be because on `variantsMap` (which I'm not using) they have duplicates on `marker` and `selection` variants.

I think it's safe to assume they're strictly progressive and the XOR operation wouldn't be impacted by generating BitVectors on the fly.

> Another piece of optimisation. Use `smallvec` instead of a `Vector`. The vector tends to save info inside the heap, but **if you know** the number of the variants like in this case, `smallvec` will save the data in the stack.

Will try to use `smallvec` and `lazy_static` as well, I don't expect any arguments to be passed to the highlighted function.

Reply:
> unless you override the whole default order configuration

We can disregard this. AFAIK, this API doesn't exist anymore. The only way to alter the order in any way is through the `before` option in plugins, but the tailwind team confirmed to me that this is a legacy API and it's likely going away soon. Copying my question and their answer here:

Q: Are utilities, components and variants added by (non-built-in) plugins ALWAYS indexed after the built-in ones?

A: Almost. The general answer is yes. There's a before option for variants which allows you to register it before another one but we consider it a legacy thing.